### PR TITLE
Hlint suggestions

### PR DESCRIPTION
**Pull request by jacereda**

### CODE REVIEW

**Comment by ndmitchell** [comment]
They look mostly good, a few I'd prefer were slightly tweaked. I'm going to go through commenting on the places that want tweaking (mostly so I have it written down for me), then merge, then fix them up myself.

File: src/Development/Make/All.hs [hunk 1]
```diff
@@ -42,7 +44,7 @@ data Ruler = Ruler
 
 eval :: Env -> Makefile -> IO [Ruler]
 eval env (Makefile xs) = do
-    (rs, env) <- runStateT (fmap concat $ mapM f xs) env
+    (rs, env) <- runStateT (concat <$> mapM f xs) env
```
Comment (ndmitchell) [inline src/Development/Make/All.hs]: Should be concatMapM

File: src/Development/Make/All.hs [hunk 2]
```diff
@@ -108,7 +110,7 @@ makePattern pat v = case break (== '%') pat of
                       else Nothing
         where rest = length v - (length pre + length post)
               subs = take rest $ drop (length pre) v
-    otherwise -> if pat == v then Just id else Nothing
+    _ -> if pat == v then Just id else Nothing
```
Comment (ndmitchell) [inline src/Development/Make/All.hs]: Eek! That one is terrible in the original, very ugly - thanks for the fix.

File: src/Development/Ninja/Lexer.hs [hunk 3]
```diff
@@ -137,22 +141,30 @@ lexBind c_x | (c,x) <- list0 c_x = case c of
...
     | (files,x) <- lexxExprs False x
     = LexDefault files : lexerLoop x
 
-lexRule x = lexxName LexRule x
-lexPool x = lexxName LexPool x
-lexInclude x = lexxFile LexInclude x
-lexSubninja x = lexxFile LexSubninja x
-lexDefine x = lexxBind LexDefine x
-
+lexRule :: Str0 -> [Lexeme]
```
Comment (ndmitchell) [inline src/Development/Ninja/Lexer.hs]: All should share one decl, e.g. `lexRule, lexPool, lexInclude :: Str0 -> [Lexeme]`

File: src/Development/Shake/Core.hs [hunk 4]
```diff
@@ -503,7 +503,7 @@ abbreviate [] = id
 abbreviate abbrev = f
     where
         -- order so longer appreviations are preferred
-        ordAbbrev = sortBy (flip (compare `on` length . fst)) abbrev
+        ordAbbrev = sortBy (flip compare `on` (length . fst)) abbrev
```
Comment (ndmitchell) [inline src/Development/Shake/Core.hs]: Should be sortOn

File: src/Development/Shake/Demo.hs [hunk 5]
```diff
@@ -63,7 +65,7 @@ demo auto = do
     createDirectoryIfMissing True dir
     forM_ ["Build.hs","main.c","constants.c","constants.h","build" <.> if isWindows then "bat" else "sh"] $ \file ->
         copyFile (manual </> file) (dir </> file)
-    when (not isWindows) $ do
+    unless isWindows $ do
          p <- getPermissions $ dir </> "build.sh"
```
Comment (ndmitchell) [inline src/Development/Shake/Demo.hs]: In general, I often find when reads better. Not a big deal here, but not a rule I like to apply universally.

File: src/Development/Shake/FileInfo.hs [hunk 6]
```diff
@@ -1,4 +1,4 @@
-{-# LANGUAGE GeneralizedNewtypeDeriving, DeriveDataTypeable, CPP, ForeignFunctionInterface #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving, DeriveDataTypeable, CPP #-}
```
Comment (ndmitchell) [inline src/Development/Shake/FileInfo.hs]: Isn't this wrong on Windows? Where I do FFI in the ifdef? I think this needs fixing before merging.

File: src/Development/Shake/FilePattern.hs [hunk 7]
```diff
@@ -184,7 +184,7 @@ compatible (x:xs) = all ((==) (specials x) . specials) xs
 
 -- | Extract the items that match the wildcards. The pair must match with '?=='.
 extract :: FilePattern -> FilePath -> [String]
-extract p@(parse -> pat) = \x -> case match pat (split isPathSeparator x) of
+extract p@(parse -> pat) x = case match pat (split isPathSeparator x) of
```
Comment (ndmitchell) [inline src/Development/Shake/FilePattern.hs]: This is very deliberate. Note that if you call `map (extract p) xs` then in the original formulation you only call parse once, which is a performance win. I need to check that GHC really does generate different code for both, and I suspect HLint told you about this, so I need to fix hlint.

File: src/Development/Shake/Progress.hs [hunk 8]
```diff
@@ -1,5 +1,4 @@
-{-# LANGUAGE DeriveDataTypeable, RecordWildCards, CPP, ForeignFunctionInterface #-}
-{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE DeriveDataTypeable, RecordWildCards, CPP, ViewPatterns #-}
```
Comment (ndmitchell) [inline src/Development/Shake/Progress.hs]: Again, a Windows-only FFI is in this module.

File: src/Development/Shake/Progress.hs [hunk 9]
```diff
@@ -299,7 +298,7 @@ jsonObject xs = "{" ++ intercalate ", " [show a ++ ":" ++ b | (a,b) <- xs] ++ "}
 xterm :: Bool
 xterm = unsafePerformIO $
     -- Terminal.app uses "xterm-256color" as its env variable
-    catch_ (fmap ("xterm" `isPrefixOf`) $ getEnv "TERM") $
+    catch_ (("xterm" `isPrefixOf`) <$> getEnv "TERM") $
```
Comment (ndmitchell) [inline src/Development/Shake/Progress.hs]: It's now got redundant brackets - you can flip the `isPrefixOf` to `isPrefixOf "xterm"`.

File: src/Development/Shake/Rules/Directory.hs [hunk 10]
```diff
@@ -108,29 +107,29 @@ instance Binary GetDirectoryQ where
 
 
 instance Rule DoesFileExistQ DoesFileExistA where
-    storedValue _ (DoesFileExistQ x) = fmap (Just . DoesFileExistA) $ IO.doesFileExist x
+    storedValue _ (DoesFileExistQ x) = (Just . DoesFileExistA) <$> IO.doesFileExist x
```
Comment (ndmitchell) [inline src/Development/Shake/Rules/Directory.hs]: Are the brackets around `(Just . DoesFileExistA)` necessary?

File: src/Development/Shake/Rules/Directory.hs [hunk 11]
```diff
@@ -269,9 +268,9 @@ removeFiles dir pat = do
         f dir2 = do
             xs <- fmap (map (dir2 </>)) $ contents $ dir </> dir2
             (dirs,files) <- partitionM (\x -> IO.doesDirectoryExist $ dir </> x) xs
-            noDirs <- fmap and $ mapM f dirs
+            noDirs <- and <$> mapM f dirs
```
Comment (ndmitchell) [inline src/Development/Shake/Rules/Directory.hs]: Should be allM

File: src/Test/FilePath.hs [hunk 12]
```diff
@@ -73,7 +73,7 @@ test build obj = do
                  ,not $ "//" `isInfixOf` noDrive y
                  ,".." `notElem` dropWhile (== "..") (dropWhile (\x -> all isPathSeparator x || isDrive x) $ splitDirectories y)
                  ,norm y == y]
-        in if and ps then True else error $ show (x, y, ps)
+        in and ps || error (show (x, y, ps))
```
Comment (ndmitchell) [inline src/Test/FilePath.hs]: I find this a bit unclear, relying on the || to hide the error.

File: src/Test/Self.hs [hunk 13]
```diff
@@ -48,7 +50,7 @@ main = shaken noTest $ \args obj -> do
         dep <- readFileLines $ out -<.> "dep"
         let xs = map (obj . moduleToFile "deps") dep
         need xs
-        ds <- fmap (nubOrd . sort . (++) dep . concat) $ mapM readFileLines xs
+        ds <- (nubOrd . sort . (++) dep . concat) <$> mapM readFileLines xs
```
Comment (ndmitchell) [inline src/Test/Self.hs]: Should be a concatMapM here as well.

File: src/Development/Shake/FilePattern.hs (hunk 7 above)
Comment (jacereda) [inline src/Development/Shake/FilePattern.hs]: If that's the case I really need to revise my beliefs on how haskell works under the hood. I thought `f x y =` is just syntactic sugar for `f x = \y ->`

Comment (ndmitchell) [inline src/Development/Shake/FilePattern.hs]: See https://github.com/ndmitchell/blogs/blob/master/hlint-view-patterns.md - it was confusing to me too, so I benchmarked on the train and wrote a blog post (that's the draft version, I'll finish it off tonight with any luck).

**Comment by ndmitchell** [comment]
I'm just merging now.

**Comment by ndmitchell** [comment]
All done, plus I fixed up all the notes I made (some fixes, some additional refactorings that were now visible).

**Comment by ndmitchell** [comment]
Thanks for taking the time to make all these improvements, a nice set of changes.

**Pull request merged**